plugins/processors/reversedns/reversedns.go

```diff
@@ -0,0 +1,193 @@
+		# get the ip from the field "source_ip", and put the result in the field "source_name"
+		field = "source_ip"
+		dest = "source_name"
+
+	[[processors.reversedns.lookup]]
+		# get the ip from the tag "destination_ip", and put the result in the tag 
+		# "destination_name".
+		tag = "destination_ip"
+		dest = "destination_name"
+
+		# If you would prefer destination_name to be a field you can use a subsequent 
+		# converter like so:
+		#   [[processors.converter.tags]]
+		#     string = ["destination_name"]
+		#     order = 2 # orders are necessary with multiple processors when order matters
+
+`
+
+type lookupEntry struct {
+	Tag   string `toml:"tag"`
+	Field string `toml:"field"`
+	Dest  string `toml:"dest"`
+}
+
+type ReverseDNS struct {
+	exitWg          *sync.WaitGroup
+	reverseDNSCache *ReverseDNSCache
+
+	Lookups            []lookupEntry `toml:"lookup"`
+	CacheTTL           time.Duration `toml:"cache_ttl"`
+	LookupTimeout      time.Duration `toml:"lookup_timeout"`
+	MaxParallelLookups int           `toml:"max_parallel_lookups"`
+}
+
+func (r *ReverseDNS) SampleConfig() string {
+	return sampleConfig
+}
+
+func (r *ReverseDNS) Description() string {
+	return "ReverseDNS "
+}
+
+type futureMetric func() telegraf.Metric
+
+func (r *ReverseDNS) Start(acc telegraf.StreamingAccumulator) error {
+	r.exitWg.Add(2)
+
+	r.reverseDNSCache = NewReverseDNSCache(r.CacheTTL, r.LookupTimeout)
+
+	orderedOutChannel := make(chan futureMetric, 10000)
+
+	go r.metricReader(acc, orderedOutChannel)
+	go r.orderedWriter(acc, orderedOutChannel)
+
+	r.exitWg.Wait()
+	return nil
+}
+
+func (r *ReverseDNS) Stop() {
+	r.exitWg.Wait()
+}
+
+func (r *ReverseDNS) metricReader(acc telegraf.StreamingAccumulator, outChannel chan<- futureMetric) {
+	defer r.exitWg.Done()
+
+	for {
+		m := acc.GetNextMetric()
+		if m == nil {
+			if acc.IsStreamClosed() {
+				close(outChannel)
+				return
+			}
+		}
+
+		outChannel <- func() telegraf.Metric {
+			workChan := make(chan telegraf.Metric)
+
+			go func(m telegraf.Metric, workChan chan telegraf.Metric) {
+				// todo: these are not in parallel. Does that matter? more promises?
+				for _, lookup := range r.Lookups {
+					if len(lookup.Field) > 0 {
+						if ipField, ok := m.GetField(lookup.Field); ok {
+							if ip, ok := ipField.(string); ok {
+								m.AddField(lookup.Dest, first(r.reverseDNSCache.Lookup(ip)))
+							}
+						}
+					}
+					if len(lookup.Tag) > 0 {
+						if ipTag, ok := m.GetTag(lookup.Tag); ok {
+							m.AddTag(lookup.Dest, first(r.reverseDNSCache.Lookup(ipTag)))
+						}
+					}
+				}
+				workChan <- m
+			}(m, workChan)
+
+			finishedMetric := <-workChan
+			return finishedMetric
+		}
+
+	}
+
+}
+
+func first(s []string) string {
+	if len(s) == 0 {
+		return ""
+	}
+	return s[0]
+}
+
+func (r *ReverseDNS) orderedWriter(acc telegraf.StreamingAccumulator, outChannel <-chan futureMetric) {
+	defer r.exitWg.Done()
+
+	for futureMetricFunc := range outChannel {
+		acc.PassMetric(futureMetricFunc())
```

Comment:
I'm confused here, all metrics are processed one at a time?  I'm okay with this as we now have every processor running concurrently and there is a tiny buffered chan between each of them, but it seems like we could do all of this with the existing Processor interface and some small changes to the RunningProcessor.

Let's talk offline about how much concurrency we need.  We don't need it at on every axis and limiting it will simplify things quite a bit.

Reply:
🤦 no just me being stupid. fixed will push.